sentry_sdk/integrations/asgi.py

```diff
@@ -0,0 +1,112 @@
+"""
+An ASGI middleware.
+
+Based on Tom Christie's `sentry-asgi <https://github.com/encode/sentry-asgi>`_.
+"""
+
+import functools
+import urllib
+
+from sentry_sdk._types import MYPY
+from sentry_sdk.hub import Hub, _should_send_default_pii
+from sentry_sdk.integrations._wsgi_common import _filter_headers
+from sentry_sdk.utils import transaction_from_function
+
+if MYPY:
+    from typing import Dict
+
+
+class SentryAsgiMiddleware:
+    __slots__ = ("app",)
+
+    def __init__(self, app):
+        self.app = app
+
+    def __call__(self, scope, receive=None, send=None):
+        if receive is None or send is None:
+
+            async def run_asgi2(receive, send):
+                return await self._run_app(
+                    scope, lambda: self.app(scope)(receive, send)
```

Comment:
@tomchristie does this look legit to you? Am I supposed to call `app(scope)` eagerly?